README.md
# fetch-vaa


This repository is no longer maintained.

The code is moved to MET Norways internal gitlab at:
https://gitlab.met.no/met/mapp/tedx/fetch-vaa


A tool to fetch Volcanic Ash Advisory messages from various websites.




Changelog
metno-fetch-vaa changelog
-------------------------

0.9.6 (2015-12-08)
Added a simple check for HTML tags in files to ensure that HTML is not
accidentally wrapped in <pre> tags.

0.9.5 (2015-12-07)
Replaced the URL for the file referenced in the TestFetcher class with one
that is publicly available.

0.9.4 (2015-10-09)
Made the fetch_vaac.py scripts's window stay on top of all others by using
a window flag that hints to the window manager that this is the desired
behaviour.
Ensured that the output directory exists and tries to create it if not.

0.9.3 (2015-08-31)
Fixed breakage of the fetch_vaac.py script by updating the use of attributes
to match that used in the underlying module.
Fixed a couple of object referencing issues in fetch_vaac.py.

0.9.2 (2015-08-11)
Enabled double clicking on list items to open the message window.

0.9.1 (2015-06-30)
Changed the file name format used for messages from Toulouse VAAC.

0.9.0 (2015-05-22)
Changed the application to use a common metno_fetch_vaa module.
Included a script and module to support integration with Ted.

0.8.0 (2015-02-20)
Added an About menu entry.

0.7.9 (2015-02-19)
Updated support for London VAAC.

0.7.8 (2014-12-15)
Fixed handling of local files, using their file names instead of their full
paths to avoid problems with storing files with derived names in the .vaac
directory.


tools/test_plots.py
#!/usr/bin/env python

# Copyright (C) 2014 MET Norway (met.no)
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA

import commands, datetime, glob, os, sys
from xml.etree import ElementTree

template = """# Created by %(program)s on %(date time)s

output = PNG

colour = COLOUR
buffersize = %(width)ix%(height)i
# papersize = A4
orientation = PORTRAIT
filename = %(output file)s
setupfile = %(setup file)s

MULTIPLE.PLOTS = %(rows)i,%(columns)i,5,5"""

plot_template = """
settime = %(plot time)s
PLOTCELL = %(row)i,%(column)i

PLOT
MAP backcolour=white map=Gshhs-Auto contour=on cont.colour=black cont.linewidth=1 cont.linetype=solid cont.zorder=1 land=on land.colour=flesh land.zorder=0 lon=off lat=off frame=off
AREA name=Europa
LABEL text="Date: %%Y-%%m-%%d %%H:%%M:%%S" valign=top fcolour=220:220:220:160 fontsize=12
#LABEL data
DRAWING file=%(kml file)s
ENDPLOT
"""

def create_input_files(setup_file, directory):

    for path in glob.glob(os.path.join(directory, "*.kml")):
    
        path = os.path.abspath(path)
        stem, suffix = os.path.splitext(path)
        input_file = stem + os.extsep + "input"
        output_file = stem + os.extsep + "png"
        
        tree = ElementTree.parse(path)
        root = tree.getroot()
        times = map(lambda element: element.text, root.findall(".//{http://www.opengis.net/kml/2.2}begin"))
        
        rows = (len(times) + 1)/2
        columns = min(len(times), 2)

        width = 400 * columns
        height = 400 * rows
        
        details = {
            "program": sys.argv[0],
            "date time": datetime.datetime.now().isoformat(" "),
            "width": width,
            "height": height,
            "output file": output_file,
            "setup file": setup_file,
            "rows": rows,
            "columns": columns
            }
        
        f = open(input_file, "w")
        f.write(template % details)
        
        row = 0
        column = 0
        
        for time in times:
        
            details = {
                "plot time": time,
                "row": row,
                "column": column,
                "kml file": path
                }
            f.write(plot_template % details)
            column = (column + 1) % 2
            if column == 0:
                row += 1
        
        f.close()

def run_bdiana(bdiana, directory):

    for input_file in glob.glob(os.path.join(directory, "*.input")):

        if os.system(commands.mkarg(bdiana) + " -i " + commands.mkarg(input_file)) != 0:
            break


if __name__ == "__main__":

    if len(sys.argv) != 4:
        sys.stderr.write("Usage: %s <bdiana executable> <setup file> <directory containing KML files>\n" % sys.argv[0])
        sys.exit(1)
    
    bdiana = sys.argv[1]
    setup_file = sys.argv[2]
    directory = sys.argv[3]

    create_input_files(setup_file, directory)
    run_bdiana(bdiana, directory)
    sys.exit()



fetch_vaac.py
#!/usr/bin/env python

# Copyright (C) 2015 MET Norway (met.no)
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA

from PyQt4 import QtCore, QtGui
import sys
import os
import urllib2
import subprocess
import selectVaac
import metno_fetch_vaa

try:
    _fromUtf8 = QtCore.QString.fromUtf8
except AttributeError:
    _fromUtf8 = lambda s: s



class MainDialog(QtGui.QDialog, selectVaac.Ui_Dialog):
    def __init__(self,fetchers,output_dir, parent=None):
        super(MainDialog,self).__init__(parent)
        self.setupUi(self)
        self.fetchers = fetchers
        self.output_dir = output_dir
        names = self.fetchers.keys()
        names.sort()
        vaacs=QtCore.QStringList(names)
        vaacs.insert(0,"Select VAAC")
        self.comboBox.insertItems(1,vaacs)
        self.comboBox.currentIndexChanged[QtCore.QString].connect(self.updateList)
        self.vaaList.currentItemChanged.connect(self.vaaListItemChanged)
        self.pushButton.clicked.connect(self.showVAACmessage)
        self.vaaList.doubleClicked.connect(self.showVAACmessage)

        self.showVAAC = QtGui.QTextBrowser()

        self.showVAAC.setGeometry(10,10,460,380)

    def updateList(self, vaac):
        QtGui.QApplication.setOverrideCursor(QtCore.Qt.WaitCursor)
        self.vaaList.clear()
        self.fetchers[str(vaac)].fetch(self.vaaList,self.output_dir)

        for i in range(self.vaaList.count()):
            item = self.vaaList.item(i)
            item.setData(QtCore.Qt.CheckStateRole, QtCore.QVariant())
            item.setFlags(QtCore.Qt.ItemIsSelectable| QtCore.Qt.ItemIsEnabled )

        QtGui.QApplication.restoreOverrideCursor()

    def accept(self):

        self.convertAdvisories()
        self.printVAACmessage()
        super(MainDialog,self).accept()

    def reject(self):

        self.printVAACmessage()

        super(MainDialog,self).reject()

    def printVAACmessage(self):
        if self.showVAAC.isVisible():
            row = self.vaaList.currentRow()
            item = self.vaaList.item(row)

            if not item.content:
                item.content = urllib2.urlopen(item.url).read()
            print item.text()
            print item.content
            geom = self.showVAAC.geometry()
            print "geom:", geom.x(), geom.y(), geom.width(), geom.height()
            self.showVAAC.close()

    def vaaListItemChanged(self):
        if self.showVAAC.isVisible():
            self.showVAACmessage()

    def showVAACmessage(self):
        row = self.vaaList.currentRow()
        item = self.vaaList.item(row)

        if not item.content:
            item.content = urllib2.urlopen(item.url).read()

        self.showVAAC.setText(item.content)
        self.showVAAC.setWindowTitle(item.text())
        self.showVAAC.show()

    def convertAdvisories(self):
        QtGui.QApplication.setOverrideCursor(QtCore.Qt.WaitCursor)

        kml_files = []
        failed_files = []
        self.workLog = {}

        row = self.vaaList.currentRow()
        if row == -1:
            return

        item = self.vaaList.item(row)
        url = item.url
        file_name = item.filename

        self.output_dir = os.path.abspath(self.output_dir)
        vaa_file = os.path.join(self.output_dir, file_name)
        if vaa_file.endswith(".html"):
            kml_file = file_name.replace(".html", ".kml")
        else:
            kml_file = file_name + ".kml"

        kml_file = os.path.join(self.output_dir, kml_file)
        message = item.text()
        if  os.path.exists(kml_file):
            message+=" already exists in " + kml_file
        else:

            if not item.content:
                vaa_content = item.content = urllib2.urlopen(url).read()
            else:
                vaa_content = item.content

            # Wrap any non-HTML content in a <pre> element.
            if not vaa_file.endswith(".html"):
                vaa_content = "<pre>\n" + vaa_content + "\n</pre>\n"
                vaa_file += ".html"

            open(vaa_file, "w").write(vaa_content)

            QtGui.QApplication.processEvents()


            # Convert the message in the HTML file to a KML file.
            s = subprocess.Popen(["/usr/bin/metno-vaa-kml", vaa_file],
                             stdout=subprocess.PIPE, stderr=subprocess.STDOUT)

            message = item.text()

            if s.wait() != 0:
                failed_files.append(vaa_file)
                item.setIcon(QtGui.QApplication.style().standardIcon(QtGui.QStyle.SP_MessageBoxWarning))
                message += " conversion failed."
            else:
                # Remove the HTML file.
                os.remove(vaa_file)
                kml_files.append(kml_file)
                item.setText(item.text() + " " + QtGui.QApplication.translate("Fetcher", "(converted)"))
                message += " converted. File available in " + kml_file

        print kml_file;

        QtGui.QApplication.restoreOverrideCursor()
        QtGui.QMessageBox.information(self, "VAAC conversion", message)

if __name__ == "__main__":


    if len(sys.argv) < 2:
        sys.stderr.write("Usage: fetch_vaac.py <directory for files>\n")
        sys.exit(1)

    output_dir = sys.argv[1]

    if not os.path.exists(output_dir):
        try:
            os.mkdir(output_dir)
        except OSError:
            sys.stderr.write("Failed to create output directory: '%s'\n" % output_dir)
            sys.exit(1)

    fetchers = {u"London VAAC": metno_fetch_vaa.LondonFetcher(), u"Toulouse VAAC": metno_fetch_vaa.ToulouseFetcher(), u"Test VAAC": metno_fetch_vaa.TestFetcher()}

    app = QtGui.QApplication(sys.argv)
    form = MainDialog(fetchers,output_dir)
    form.setWindowFlags(form.windowFlags() | QtCore.Qt.WindowStaysOnTopHint)
    form.show()

    sys.exit(app.exec_())


metno_fetch_vaa.py
#!/usr/bin/env python

# Copyright (C) 2014 MET Norway (met.no)
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA

import commands, datetime, HTMLParser, os, subprocess, sys, urllib2, urlparse

from PyQt4.QtCore import *
from PyQt4.QtGui import *

__version__ = "0.9.4"

checked_dict = {False: Qt.Unchecked, True: Qt.Checked}

class Settings(QSettings):

    """Convenience class to help read values from settings files as Python datatypes.
    """

    def __init__(self, organisation, product):

        QSettings.__init__(self, organisation, product)
    
    def value(self, key, default = QVariant()):
    
        """Reads the value from the settings file that corresponds to the given key,
        with the type defined by the default value. If the key is not defined in the
        settings file then the default value is returned instead.
        """
        value = QSettings.value(self, key, default)
        if type(default) == int:
            return value.toInt()[0]
        elif type(default) == bool:
            return value.toBool()
        elif type(default) == float:
            return value.toFloat()[0]
        elif type(default) == str:
            return str(value.toString())
        elif type(default) == unicode:
            return unicode(value.toString())
        elif type(default) == list:
            if type(default[0]) == int:
                return map(lambda v: v.toInt()[0], value.toList())
        else:
            return value
        
        raise ValueError, "Unknown value type for key '%s': %s" % (key, type(default))


class Parser(HTMLParser.HTMLParser):

    def __init__(self):
    
        HTMLParser.HTMLParser.__init__(self)
        self.href = ""
        self.text = ""
        self.table_row = []
        self.tags = []
    
    def feed(self, data):

        self.anchors = []
        HTMLParser.HTMLParser.feed(self, data)
    
    def handle_starttag(self, tag, attrs):
    
        self.tags.append(tag)
        
        if tag == "a":
            d = dict(attrs)
            try:
                self.href = d["href"]
            except KeyError:
                pass
        elif tag == "tr":
            self.table_row = []
        
        self.text = ""

    def handle_data(self, data):
    
        self.text += data
    
    def handle_endtag(self, tag):

        if tag == "a":
            self.anchors.append((self.href, self.text, self.table_row))
            self.href = ""
        elif tag == "td":
            self.table_row.append(self.text)
        
        # Discard any non-matching end tags.
        while self.tags:
            if self.tags.pop() == tag:
                break


class Fetcher:

    showBusy = True
    showInMenu = True
    defaultFlags = Qt.ItemIsEnabled | Qt.ItemIsUserCheckable

    def hasExistingFile(self, output_dir, href):

        file_name = href.split("/")[-1]
         
        vaa_file = os.path.join(output_dir, file_name)
        if vaa_file.endswith(".html"):
            kml_file = file_name.replace(".html", ".kml")
        else:
            kml_file = file_name + ".kml"
        
        return os.path.exists(os.path.join(output_dir, kml_file))


class ToulouseFetcher(Fetcher):

    url = "http://www.meteo.fr/vaac/evaa.html"
    number_to_fetch = 10
    returns_html = True

    def fetch(self, vaaList, output_dir):
    
        "Reads the messages available from the URL for the current VAA centre."

        html = urllib2.urlopen(self.url).read()
        p = Parser()
        p.feed(html)
        p.close()
        
        count = 0

        for href, text, table_text in p.anchors:

            if text.endswith("UTC"):
            
                # The date is encoded in the URL for the advisory.
                info = href.split(".")
                date = datetime.datetime.strptime(info[-2], "%Y%m%d%H%M").strftime("%Y-%m-%d %H:%M")
                volcano = info[2].replace("_", " ")
                item = QListWidgetItem("%s (%s)" % (date, volcano))
                item.setFlags(self.defaultFlags)
                # The name to use for the locally stored file needs to be
                # in a suitable format so that Diana can read it as part of
                # a collection.
                item.filename = "toulouse." + info[-2] + ".html"
                item.url = urlparse.urljoin(self.url, href)
                item.content = None
                item.setCheckState(Qt.Unchecked)
                if self.hasExistingFile(output_dir, item.filename):
                    item.setText(item.text() + " " + QApplication.translate("Fetcher", "(converted)"))
                vaaList.addItem(item)

                count += 1
                if count == self.number_to_fetch:
                    break


class AnchorageFetcher(Fetcher):

    url = "http://vaac.arh.noaa.gov/list_vaas.php"
    number_to_fetch = 10
    returns_html = True

    def fetch(self, vaaList, output_dir):
    
        "Reads the messages available from the URL for the current VAA centre."

        html = urllib2.urlopen(self.url).read()
        p = Parser()
        p.feed(html)
        p.close()
        
        count = 0

        for href, text, table_text in p.anchors:

            if text == "X" and href.split("/")[-2] == "VAA":
            
                # The date is encoded in the associated table text.
                date = datetime.datetime.strptime(table_text[0], "%Y-%m-%d %H:%M:%S").strftime("%Y-%m-%d %H:%M:%S")
                volcano = table_text[1].replace("_", " ")
                item = QListWidgetItem("%s (%s)" % (date, volcano))
                item.setFlags(self.defaultFlags)
                item.filename = href
                item.url = urlparse.urljoin(self.url, href)
                item.content = None
                item.setCheckState(Qt.Unchecked)
                if self.hasExistingFile(output_dir, href):
                    item.setText(item.text() + " " + QApplication.translate("Fetcher", "(converted)"))
                vaaList.addItem(item)

                count += 1
                if count == self.number_to_fetch:
                    break


class LondonFetcher(Fetcher):

    # Scrape the London VAAC page on the Met Office website.
    url = "http://www.metoffice.gov.uk/aviation/vaac/vaacuk.html"
    # A listing of the files can be obtained from this address:
    # http://www.metoffice.gov.uk/aviation/vaac/data/?C=M;O=D
    number_to_fetch = 10
    returns_html = False

    def fetch(self, vaaList, output_dir):
    
        "Reads the messages available from the URL for the current VAA centre."

        html = urllib2.urlopen(self.url).read()
        p = Parser()
        p.feed(html)
        p.close()
        
        count = 0
        # Some message appear more than once in the table, so filter out duplicates.
        urls = set()

        for href, text, table_text in p.anchors:

            if text == "VAA" and href.endswith(".html"):
            
                # The date is encoded in the associated table text.
                # Unfortunately, it is localised, so we use the date from the
                # message itself.
                message_url = urlparse.urljoin(self.url, href)
                if message_url in urls:
                    continue

                urls.add(message_url)

                volcano, date, text = self.read_message(message_url)
                item = QListWidgetItem("%s (%s)" % (date, volcano))
                item.setFlags(self.defaultFlags)
                item.filename = "london." + date.strftime("%Y%m%d%H%M")
                item.url = message_url
                item.content = text
                item.setCheckState(Qt.Unchecked)
                if self.hasExistingFile(output_dir, item.filename):
                    item.setText(item.text() + " " + QApplication.translate("Fetcher", "(converted)"))
                vaaList.addItem(item)

                count += 1
                if count == self.number_to_fetch:
                    break

    def read_message(self, url):
    
        html = urllib2.urlopen(url).read()

        # The London VAAC currently shows advisories in HTML pages.
        # We just extract what we can find.
        at = html.find("VA ADVISORY")
        if at == -1:
            return
        
        start = html.rfind("<p>", 0, at)
        if start == -1:
            return
        
        start = html.find("<br>", start)
        lines = html[start:].split("\n")
        
        text = ""
        date = datetime.datetime.now()
        volcano = "Unknown"

        for line in lines:

            line = line.strip()
            if not line or line == "<br>":
                continue
            
            text += line + "\n"
            if line.startswith("DTG:"):
                # The date is encoded in the advisory.
                date_text = line[4:].lstrip()
                date = datetime.datetime.strptime(date_text, "%Y%m%d/%H%MZ")
            
            if line.startswith("VOLCANO:"):
                # The volcano is encoded in the advisory.
                volcano = line[8:].lstrip()

            if line == "=":
                break
        
        return volcano, date, text


class LocalFileFetcher(Fetcher):

    returns_html = False
    showBusy = False
    showInMenu = False

    def fetch(self, vaaList, output_dir):
    
        fileName = QFileDialog.getOpenFileName(None, QApplication.translate("LocalFileFetcher", "Open VAA File"))
        
        if fileName.isEmpty():
            return
        
        fileName = unicode(fileName)
        
        vaaList.clear()
        item = QListWidgetItem(fileName)
        item.setFlags(self.defaultFlags)
        item.filename = os.path.split(fileName)[1]
        item.url = urlparse.urljoin("file://", fileName)
        item.content = None
        item.setCheckState(Qt.Unchecked)
        vaaList.addItem(item)


class TestFetcher(Fetcher):

    url = "https://github.com/metno/fetch-vaa/raw/master/files/london-201511101500.vaa.txt"
    number_to_fetch = 1
    returns_html = True
    
    def fetch(self, vaaList, output_dir):
    
        "Reads the messages available from the URL for the current VAA centre."

        text = urllib2.urlopen(self.url).read()
        
        date = datetime.datetime.now()
        volcano = "Unknown"

        lines = text.split("\n")
        for line in lines:

            if line.startswith("DTG:"):
                # The date is encoded in the advisory.
                date_text = line[4:].lstrip()
                date = datetime.datetime.strptime(date_text, "%Y%m%d/%H%MZ")
        
            if line.startswith("VOLCANO:"):
                # The volcano is encoded in the advisory.
                volcano = line[8:].lstrip()
        
        item = QListWidgetItem("%s (%s)" % (date.strftime("%Y-%m-%d %H:%M:%S"), volcano))
        item.setFlags(self.defaultFlags)
        # Use a different name for the path instead of the path of the page on the site.
        item.filename = "test." + date.strftime("%Y%m%d%H%M")
        # Store the original location.
        item.url = self.url
        # We have already obtained the content.
        item.content = text
        item.setCheckState(Qt.Unchecked)
        if self.hasExistingFile(output_dir, item.filename):
            item.setText(item.text() + " " + QApplication.translate("Fetcher", "(converted)"))
        vaaList.addItem(item)


class EditDialog(QDialog):

    def __init__(self, content, parent = None):

        QDialog.__init__(self, parent)

        self.textEdit = QPlainTextEdit()
        self.textEdit.setPlainText(content)
        
        buttonBox = QDialogButtonBox()
        buttonBox.addButton(QDialogButtonBox.Ok)
        buttonBox.addButton(QDialogButtonBox.Cancel)
        
        buttonBox.accepted.connect(self.accept)
        buttonBox.rejected.connect(self.reject)

        layout = QVBoxLayout(self)
        layout.addWidget(self.textEdit)
        layout.addWidget(buttonBox)

        self.setWindowTitle(self.tr("Edit Message"))


class Window(QMainWindow):

    def __init__(self, fetchers):

        QMainWindow.__init__(self)

        self.fetchers = fetchers
        self.settings = Settings("met.no", "metno-fetch-vaa")
        
        self.output_dir = self.settings.value("work directory",
                          os.path.join(os.getenv("HOME"), ".vaac"))
        self.workLog = {}
        
        contentWidget = QWidget()
        layout = QGridLayout(contentWidget)
        
        fileMenu = self.menuBar().addMenu(self.tr("&File"))
        newFileAction = fileMenu.addAction(self.tr("&New File..."), self.newFile,
            QKeySequence.New)
        openFileAction = fileMenu.addAction(self.tr("&Open File..."), self.fetchAdvisories,
            QKeySequence.Open)
        openFileAction.name = u"Local file"

        fileMenu.addSeparator()

        # Create a list of available VAA centres in the menu.
        names = self.fetchers.keys()
        names.sort()

        for name in names:
            if self.fetchers[name].showInMenu:
                action = fileMenu.addAction(name, self.fetchAdvisories)
                action.name = name
        
        fileMenu.addSeparator()
        fileMenu.addAction(self.tr("E&xit"), self.close, QKeySequence(QKeySequence.Quit))

        # Add a Help menu with about and documentation entries.
        helpMenu = self.menuBar().addMenu(self.tr("&Help"))
        aboutAction = helpMenu.addAction(self.tr("&About..."), self.about)
        
        # Create a list of downloaded advisories.
        self.vaaList = QListWidget()
        layout.addWidget(self.vaaList, 0, 0)
        
        # Add a panel of buttons.
        buttonLayout = QHBoxLayout()

        self.editButton = QPushButton(self.tr("&Edit message"))
        self.editButton.setSizePolicy(QSizePolicy.Maximum, QSizePolicy.Fixed)
        buttonLayout.addWidget(self.editButton)
        buttonLayout.setAlignment(self.editButton, Qt.AlignHCenter)

        self.convertButton = QPushButton(self.tr("&Convert messages"))
        self.convertButton.setSizePolicy(QSizePolicy.Maximum, QSizePolicy.Fixed)
        buttonLayout.addWidget(self.convertButton)
        buttonLayout.setAlignment(self.convertButton, Qt.AlignHCenter)

        layout.addLayout(buttonLayout, 1, 0)
        
        # Ensure that the list widgets are given enough space.
        layout.setRowStretch(1, 1)

        # Add a log viewer.
        self.logViewer = QPlainTextEdit()
        self.logViewer.setReadOnly(True)
        self.showHideLogButton = QPushButton(self.tr("&Hide log"))
        self.showHideLogButton.setSizePolicy(QSizePolicy.Maximum, QSizePolicy.Fixed)

        layout.addWidget(self.logViewer, 3, 0)
        layout.addWidget(self.showHideLogButton, 4, 0)
        self.showHideLogViewer(self.settings.value("window/log", False))

        # Make connections.
        self.vaaList.currentItemChanged.connect(self.showLog)
        self.vaaList.itemActivated.connect(self.showLog)
        self.vaaList.currentItemChanged.connect(self.updateButtons)
        self.vaaList.itemChanged.connect(self.updateButtons)
        self.vaaList.itemActivated.connect(self.updateButtons)
        self.editButton.clicked.connect(self.editMessage)
        self.convertButton.clicked.connect(self.convertAdvisories)
        self.showHideLogButton.clicked.connect(self.showHideLogViewer)
        
        self.updateButtons()

        self.setCentralWidget(contentWidget)
        self.setWindowTitle(self.tr("Fetch Volcanic Ash Advisories"))
        if self.settings.contains("window/geometry"):
            self.restoreGeometry(self.settings.value("window/geometry").toByteArray())
        else:
            self.resize(640, 480)
    
    def about(self):

        QMessageBox.about(self, self.tr("About this program"),
            self.tr("<qt>Fetches Volcanic Ash Advisory (VAA) messages from certain "
                    "Volcanic Ash Advisory Centres (VAAC) and converts them to "
                    "Keyhole Markup Language (KML) files for use with Diana and "
                    "Ted.<p><b>Version:</b> %1</qt>").arg(__version__))
    
    def newFile(self):
    
        # Ask for the name of the file.
        volcano, success = QInputDialog.getText(self, self.tr("New File"),
            self.tr("Volcano name:"))
        
        if success and volcano:
            volcano = unicode(volcano)
        else:
            volcano = u"Unknown"
        
        date = datetime.datetime.now().strftime("%Y%m%d%H%M%S")
        fileName = os.path.join(self.output_dir, u"%s.%s" % (volcano, date))
        
        # Create the file.
        try:
            open(fileName, "w").write("")
        except IOError:
            QMessageBox.critical(self, self.tr("Error"),
                self.tr("Failed to create an empty file to use for a new message.\n"
                        'Please consult the documentation for support.'))
            return
        
        # Add an item to the list.
        item = QListWidgetItem(fileName)
        item.setFlags(Qt.ItemIsEnabled | Qt.ItemIsSelectable)
        item.filename = fileName
        item.url = urlparse.urljoin("file://", fileName)
        item.content = None
        item.setCheckState(checked_dict[False])
        self.vaaList.addItem(item)

        self.updateButtons()
        self.vaaList.setCurrentItem(item)
        self.editButton.animateClick()

    def fetchAdvisories(self):
    
        self.vaaList.clear()
        name = self.sender().name
        fetcher = self.fetchers[name]
        
        # Create the output directory if it does not already exist.
        if not os.path.exists(self.output_dir):
            os.system("mkdir -p " + commands.mkarg(self.output_dir))
        
        if fetcher.showBusy:
            QApplication.setOverrideCursor(Qt.WaitCursor)
        
        fetcher.fetch(self.vaaList, self.output_dir)

        self.workLog = {}
        self.updateButtons()
        
        if fetcher.showBusy:
            QApplication.restoreOverrideCursor()
            
            if self.vaaList.count() == 0:
                QMessageBox.information(self, self.tr("Fetching from %1").arg(name),
                    self.tr("No new messages available from %1.").arg(name))

    def updateButtons(self):
    
        yet_to_convert = False

        for i in range(self.vaaList.count()):
            item = self.vaaList.item(i)
            if item.checkState() == Qt.Checked and \
               not os.path.exists(os.path.join(self.output_dir, item.filename)):
               yet_to_convert = True
               break

        self.convertButton.setEnabled(yet_to_convert)
        editable = self.vaaList.count() > 0 and self.vaaList.currentRow() != -1
        self.editButton.setEnabled(editable)
    
    def convertAdvisories(self):
    
        QApplication.setOverrideCursor(Qt.WaitCursor)
        
        kml_files = []
        failed_files = []
        self.workLog = {}
        
        for i in range(self.vaaList.count()):
        
            item = self.vaaList.item(i)
            if not item.checkState() == Qt.Checked:
                continue
            
            href = item.filename
            url = item.url
            
            file_name = href.split("/")[-1]
            
            vaa_file = os.path.join(self.output_dir, file_name)
            if vaa_file.endswith(".html"):
                kml_file = file_name.replace(".html", ".kml")
            else:
                kml_file = file_name + ".kml"
            
            kml_file = os.path.join(self.output_dir, kml_file)
            if os.path.exists(kml_file):
                continue

            if not item.content:
                vaa_content = item.content = urllib2.urlopen(url).read()
            else:
                vaa_content = item.content

            # Wrap any non-HTML content in a <pre> element.
            if vaa_content.lstrip()[:1] != "<":
                vaa_content = "<pre>\n" + vaa_content + "\n</pre>\n"
            if not vaa_file.endswith(".html"):
                vaa_file += ".html"

            open(vaa_file, "w").write(vaa_content)
            
            QApplication.processEvents()

            # Convert the message in the HTML file to a KML file.
            s = subprocess.Popen(["/usr/bin/metno-vaa-kml", vaa_file],
                                 stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
            if s.wait() != 0:
                failed_files.append(vaa_file)
                item.setIcon(QApplication.style().standardIcon(QStyle.SP_MessageBoxWarning))
            else:
                # Remove the HTML file.
                os.remove(vaa_file)
                kml_files.append(kml_file)
                item.setText(item.text() + " " + QApplication.translate("Fetcher", "(converted)"))

            self.workLog[item.filename] = s.stdout.read()
        
        # Update the log viewer if it is already shown.
        if self.logViewer.isVisible():
            self.showLog()

        self.updateButtons()
        
        QApplication.restoreOverrideCursor()
    
    def showLog(self):
    
        item = self.vaaList.currentItem()
        if item and item.filename in self.workLog:
            text = self.workLog[item.filename]
            self.logViewer.setPlainText(text)
            self.showHideLogViewer(True)
        else:
            self.logViewer.clear()
    
    # Use a decorator to avoid receiving the signal that includes a boolean value.
    @pyqtSlot()
    def showHideLogViewer(self, show = None):
    
        if show is None:
            show = not self.logViewer.isVisible()
        
        self.logViewer.setShown(show)
        if show:
            self.showHideLogButton.setText(self.tr("&Hide log"))
        else:
            self.showHideLogButton.setText(self.tr("&Show log"))
    
    def editMessage(self):

        row = self.vaaList.currentRow()
        item = self.vaaList.item(row)

        if not item.content:
            item.content = urllib2.urlopen(item.url).read()
        
        oldContent = item.content

        editDialog = EditDialog(item.content[:], self)
        editDialog.restoreGeometry(self.settings.value("editdialog/geometry").toByteArray())

        if editDialog.exec_() == QDialog.Accepted:
            item.content = unicode(editDialog.textEdit.toPlainText())
            if oldContent != item.content:
                item.setCheckState(Qt.Unchecked)

        self.updateButtons()
    
    def closeEvent(self, event):

        self.settings.setValue("window/geometry", self.saveGeometry())
        self.settings.setValue("window/log", self.logViewer.isVisible())
        self.settings.setValue("editdialog/geometry", self.saveGeometry())
        self.settings.sync()

